ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java

```diff
@@ -326,8 +327,11 @@ private StreamInfo newStreamInfo(ByteBuf buf) {
   }
 
   static long writeTo(ByteBuf buf, DataStream stream) {
-    final WritableByteChannel channel = stream.getWritableByteChannel();
     long byteWritten = 0;
+    if (stream == null) {
```

Comment:
Why the stream could be null?

Reply:
In current test I am using BaseStateMachine, in which the stream api returns:

https://github.com/apache/incubator-ratis/blob/f91ef63809b21ed65445f4a60d8005ce0a1af8ca/ratis-server/src/main/java/org/apache/ratis/statemachine/StateMachine.java#L91

Reply:
We should override DataApi.stream(RaftClientRequest request) in the test.  Otherwise, how could it test the correctness?  We may use MultiDataStreamStateMachine in DataStreamBaseTest.

Reply:
Yeah that what I was thinking in https://issues.apache.org/jira/browse/RATIS-1130?focusedCommentId=17228841&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17228841

Sure I can move `MultiDataStreamStateMachine` to make it reusable. Do you prefer I do that in another PR (as this PR is pretty large now) or just add that refactoring in this PR?

Also adding a `NULL` is still useful. I found that without this NULL check, this code will hit a NPE and the program will hang there. If don't add this NULL check, then at least we should check the NPE and solve why program will hang.

Reply:
Another thought is, with MiniRaftCluster based test, do we still need existing test cases? Can we migrate all tests to MiniRaftCluster based?

Reply:
We need existing test case. We need both unit test and integration Test.

Reply:
I  do not think we need to add a NULL check. if NULL happen in test, we should change the test to avoid this. Please think about it.

Reply:
I will hold off if we need the `NULL` check for now because it is not related to test itself. 


Back to whether we need `MultiDataStreamStateMachine` for this MiniRaftCluster test, here is my question:

If we enable `MultiDataStreamStateMachine` for MiniRaftCluster test, what is the difference between this MiniRaftCluster test and the existing `TestDataStreamNetty`? From the test coverage MiniRaftCluster will be a super set over `TestDataStreamNetty`, why we want to keep both?

Reply:
Done